metaphor/power_bi/extractor.py

```
@@ -0,0 +1,411 @@
+        url = f"{self.API_ENDPOINT}/groups/{group_id}/reports/{report_id}"
+        return self._call_get(url, PowerBIReport)
+
+    def get_datasource(
+        self, group_id: str, dashboard_id: str
+    ) -> List[PowerBIDataSource]:
+        url = (
+            f"{self.API_ENDPOINT}/groups/{group_id}/datasets/{dashboard_id}/datasources"
+        )
+        return self._call_get(url, List[PowerBIDataSource])
+
+    def get_workspace_info(self, workspace_id: str) -> WorkspaceInfo:
+        def create_scan() -> str:
+            url = f"{self.API_ENDPOINT}/admin/workspaces/getInfo"
+            request_body = {"workspaces": [workspace_id]}
+            result = requests.post(
+                url,
+                headers=self._headers,
+                params={
+                    "datasetExpressions": True,
+                    "datasetSchema": True,
+                    "datasourceDetails": True,
+                    "getArtifactUsers": True,
+                    "lineage": True,
+                },
+                data=request_body,
+            )
+
+            return result.json()["id"]
+
+        def wait_for_scan_result(scan_id: str, max_timeout_in_secs: int = 30) -> bool:
+            url = f"{self.API_ENDPOINT}/admin/workspaces/scanStatus/{scan_id}"
+
+            waiting_time = 0
+            while True:
+                result = requests.get(url, headers=self._headers)
+                if result.status_code != 200:
+                    return False
+                if result.json()["status"] == "Succeeded":
+                    return True
+                if waiting_time >= max_timeout_in_secs:
+                    break
+                waiting_time += 1
+                sleep(1)
+            return False
+
+        scan_id = create_scan()
+        wait_for_scan_result(scan_id)
+        url = f"{self.API_ENDPOINT}/admin/workspaces/scanResult/{scan_id}"
+
+        def get_first_workspace(response: requests.Response) -> Any:
```

Add a comment why we're only interested in the first workspace?

---

Also what happens if the response doesn't contain any workspace, e.g. the wrong workspace ID was specified in the config.

---

Added a comment.
The response should be expected otherwise would fail earlier. If workspace ID is not valid or no access to the workspace, the first step will throw an exception. If something goes wrong when polling the scan status, it should throw an exception too.